runtime/runtime/src/lib.rs

```diff
@@ -1690,12 +1690,12 @@ impl Runtime {
                     }),
                 };
 
-                // For yielded promises the sender is always the receiver. We can process
-                // the receipt directly because we know it is destined for the local shard.
-                //
-                // Note that we don't invoke the prefetcher as it doesn't do anything
-                // for data receipts.
-                process_receipt(&resume_receipt, &mut state_update, &mut total)?;
+                receipt_sink.forward_or_buffer_receipt(
```

Comment:
This part is a behavior change. It addresses the following issue:

When the `yield_resume` host function is invoked successfully it produces an outgoing receipt and returns a success signal. We then guarantee that the yield callback will be executed with a non-timeout input.

Suppose that `yield_resume` is invoked successfully **and** the yield timeout is reached within the same chunk. If we apply the timeout-generated receipt immediately, the yield callback will be executed with a timeout input (this is incorrect behavior). By handling the timeout as an outgoing receipt as well, we allow the `yield_resume`-created receipt to be processed first.

Generating multiple resume receipts for the same yield is not problematic; all but the first to be processed will be safely discarded.

Note that this approach is dependent on the fact that outgoing receipts which are destined for the same shard will be applied in the order in which they are created.

Reply:
cc @jakmeier on use of `forward_or_buffer_receipt`.

Reply:
Calling `forward_or_buffer_receipt` with a receipt destined to the same shard, it will directly put it in the outgoing receipts buffer. Equivalent to `outgoing_receipts.push(receipt)` before any of the congestion control changes.

I think that is what you want to happen, to fix this corner case, right? @saketh-are 
I am not familiar enough with the yield-resume mechanic to comment on whether this is the best way of fixing it, or if it even fixes it entirely. But I can say that this will put the receipt in the outgoing receipts for execution at a future point, no sooner than one block height later. 

On a side note, if reviewers need a description of why this makes sense, my rule of thumb is that comments in the code should contain at least as generous a description, since future readers will be less familiar with the feature.